src/sqlancer/ASTBasedReducer.java

```diff
@@ -0,0 +1,470 @@
+        }
+
+        @Override
+        public void visit(CaseExpression expr) {
+            Expression switchExpr = expr.getSwitchExpression();
+            Expression elseExpr = expr.getElseExpression();
+
+            expressionReduce(expr, switchExpr, (parent, sw) -> {
+                parent.setSwitchExpression(sw);
+                updateStatements();
+            });
+            expressionReduce(expr, elseExpr, (parent, els) -> {
+                parent.setElseExpression(els);
+                updateStatements();
+            });
+
+            super.visit(expr);
+        }
+
+        @Override
+        public void visit(WhenClause whenClause) {
+            Expression when = whenClause.getWhenExpression();
+            Expression then = whenClause.getThenExpression();
+
+            expressionReduce(whenClause, when, (wc, w) -> {
+                wc.setWhenExpression(w);
+                updateStatements();
+            });
+
+            expressionReduce(whenClause, then, (wc, t) -> {
+                wc.setThenExpression(t);
+                updateStatements();
+            });
+
+            super.visit(whenClause);
+        }
+
+        @Override
+        public void visit(Parenthesis parenthesis) {
+            Expression closedExpr = parenthesis.getExpression();
+            expressionReduce(parenthesis, closedExpr, (p, s) -> {
+                p.setExpression(s);
+                updateStatements();
+            });
+            closedExpr.accept(this);
+        }
+
+        @Override
+        public void visit(Function function) {
+            super.visit(function);
+        }
+
+        @Override
+        public void visit(ExpressionList expressionList) {
+            List<Expression> expressions = expressionList.getExpressions();
+            listElementRemovingReduce(expressionList, expressions, (l, es) -> {
+                l.setExpressions(es);
+                updateStatements();
+            });
+            expressions = expressionList.getExpressions();
+            for (int i = 0; i < expressions.size(); i++) {
+                Expression expr = expressions.get(i);
+                int index = i;
+                expressionReduce(expressions, expr, (l, e) -> {
+                    l.set(index, e);
+                    updateStatements();
+                });
+            }
+            super.visit(expressionList);
+        }
+
+        @Override
+        public void visit(SelectExpressionItem selectExpressionItem) {
+            Expression expr = selectExpressionItem.getExpression();
+            expressionReduce(selectExpressionItem, expr, (item, e) -> {
+                item.setExpression(e);
+                updateStatements();
+            });
+            super.visit(selectExpressionItem);
+        }
+    };
+
+    SelectVisitorAdapter selectReducerVisitor = new SelectVisitorAdapter() {
+        private final String[] removeList = { "Limit", "Offset", "Where", "Having", "GroupBy", "Distinct",
```

Comment:
I think it would be useful to have some comments for what these three arrays mean/how they are used